Add support for syllabi

ap-1 opened the pull request:
Created a `SideNav` entry for syllabi, a section for it in `CourseCard`, and added it to the prisma schema.

I also attempted to created a syllabus controller and create routes for it in the backend, but I need someone to verify I'm doing it correctly. I also need help writing `app/api/syllabi.ts` and consuming the API properly in `CourseCard`.

The syllabus page and search functionality for it are also not yet implemented, and I will probably need help with the latter

Closes https://github.com/ScottyLabs/cmucourses/issues/201 and https://github.com/ScottyLabs/cmucourses/issues/121

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

ilong4rennes commented:
@ap-1 Hi Anish,

I've implemented syllabi functionality:
- Added `app/api/syllabi.ts` API module
- Added Course Syllabus Card to `CourseDetails.tsx`
- Added view for `/syllabi` route with basic `SyllabiSearchList.tsx` for debugging

**Issue found**: While the database contains 19,556 syllabi records, all field values are NULL (see screenshots). Since you're familiar with the course scraper, could you help diagnose why the data wasn't properly populated?

![Screenshot 2025-03-13 at 7 36 43 PM](https://github.com/user-attachments/assets/4de32ffe-dbc1-4db6-90b5-cb0f0172ede1)
![Screenshot 2025-03-13 at 7 36 14 PM](https://github.com/user-attachments/assets/104ce233-330f-48e9-bd83-8e45707197a1)
![Screenshot 2025-03-13 at 8 14 12 PM](https://github.com/user-attachments/assets/9c17a741-dfc8-49da-9f7e-3ab6f104c69a)

ap-1 commented:
woah this is great

i'm actually working on a rewrite of course-api in a separate repo, and I moved the syllabus scraper there along w a bunch of other updates---i'll upload the csv to mongo for now

do note that I removed a bunch of redundant data, including titles, in exchange for having more reliable course number/section parsing

ap-1 commented:
I think I pushed the wrong commit the first time but it should be fixed now w/ your changes and the updated schema in `schema.prisma`. I also ran `MONGODB_URI="....." bun run db-migrate`
```bash
$ prisma db push
Prisma schema loaded from schema.prisma
Datasource "db": MongoDB database "course-api" at "monorail.proxy.rlwy.net:39417"
Applying the following changes:

[-] Index `year_1_number_1_section_1_season_1`
[-] Index `text`
[+] Fulltext index `text` on ({"name":"text","department":"text","desc":"text","prereqString":"text"})
[+] Index `syllabi_year_number_section_season_idx` on ({"year":1,"number":1,"section":1,"season":1})


🚀  Your database indexes are now in sync with your Prisma schema. Done in 2.59s

✔ Generated Prisma Client (v5.22.0) to ./../../node_modules/@prisma/client in 51ms
```

ilong4rennes commented:
Added the MVP of the frontend. The link to the syllabi is now clickable from the course details, and the syllabi list view is now available, but the display does not look beautiful. I will finish this up hopefully by tomorrow or Saturday.

ap-1 commented:
nice! one thing I would note is that users would probably want the option whether to automatically open the download link or view the PDF/docx/whatever in canvas. This can be done by modifying the URL as such:

(dowload link)
https://canvas.cmu.edu/files/5177329/download?download_frd=1&verifier=5dM1dEf8IwZFQcqC1MIou46HO1kVtEzX2Ru1IYvD

(canvas link)
https://canvas.cmu.edu/files/5177329/?verifier=5dM1dEf8IwZFQcqC1MIou46HO1kVtEzX2Ru1IYvD

ap-1 commented:
I like what you did with the Course Syllabus card (the one w the placeholder "No syllabi available for this course"), I think it makes it cleaner when a course has multiple syllabi to display. Then the "Syllabi" section in in CourseCard could simply link to the Course Syllabus card on the /syllabi route, instead of providing a download link to a single syllabus (it isn't clear which one it is providing the syllabus to before it's actually clicked on)

There's a discord bot that does something similar to what I'm thinking for this card:

<img width="472" alt="image" src="https://github.com/user-attachments/assets/962e18ac-43d8-404d-98e2-b2e635c05558" />

We can do something like this w/ an actual UI and get the instructor information from looking up the course+section+season+year in the DB.

ilong4rennes commented:
> I like what you did with the Course Syllabus card (the one w the placeholder "No syllabi available for this course"), I think it makes it cleaner when a course has multiple syllabi to display. Then the "Syllabi" section in in CourseCard could simply link to the Course Syllabus card on the /syllabi route, instead of providing a download link to a single syllabus (it isn't clear which one it is providing the syllabus to before it's actually clicked on)
> 
> There's a discord bot that does something similar to what I'm thinking for this card:
> 
> <img alt="image" width="472" src="https://private-user-images.githubusercontent.com/67872951/423161679-962e18ac-43d8-404d-98e2-b2e635c05558.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIwNzk0OTgsIm5iZiI6MTc0MjA3OTE5OCwicGF0aCI6Ii82Nzg3Mjk1MS80MjMxNjE2NzktOTYyZTE4YWMtNDNkOC00MDRkLTk4ZTItYjJlNjM1YzA1NTU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE1VDIyNTMxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlN2U4OGRkNTk1OWI1NDk5M2RhZDZhY2I1NTI4NzFlMzRkMjkyYjA1YjU1NTFjMjhjMGY1ZmQyZjBlNDJmMjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.apVkQ2hccXeLuLyLogBRakA-qt3dtp9x15dBZn85fEA">
> We can do something like this w/ an actual UI and get the instructor information from looking up the course+section+season+year in the DB.

This is a good idea! Ill work on this!

ap-1 commented:
this is great work! i really appreciate all the effort you've put into this. i just have a few small nitpicks (and i apologize in advance if some of these sound a little pedantic)

1. in the course card, we have an established style for links that is used in the prereqs/coreqs/crosslisted sections. it would be great if the syllabus section used the same style.

<img width="191" alt="image" src="https://github.com/user-attachments/assets/3187aa7c-f7b1-491d-94dd-055ceaef4123" />

2. Rather than the heading being "Syllabus" and the link being "View All Syllabi," I think it would look slightly cleaner if the heading was "Syllabi" and the link was "View All"

3. In the course cards, the headings aren't styled like links, and they have no hover effects aside from `cursor-pointer`. it would be great if the syllabus cards followed the same format. Also, the course cards have a little bit of padding (margin?) under the heading, and i think the syllabus cards should have that too, just so that the table isn't right up against it

<img width="442" alt="image" src="https://github.com/user-attachments/assets/22c33872-6d7c-4ce7-914a-50789d22d060" />
<img width="425" alt="image" src="https://github.com/user-attachments/assets/01125f2d-610c-4826-ae6a-bd913e43575a" />

4. Because the dedicated course pages don't have a heading or a "View All Courses" button, I don't think the dedicated syllabus pages should have a "Syllabi for XX-XXX" heading or a "View All Syllabi" button either. The page should ideally start with the table (which contains the heading), like what the course pages do

<img width="756" alt="image" src="https://github.com/user-attachments/assets/cd94c786-d1b1-4f8b-b5b8-8b26a572755e" />
<img width="754" alt="image" src="https://github.com/user-attachments/assets/48cda2c4-cd09-4b35-b945-0a6c860f2063" />

5. I think these sections would also benefit from using the same link style as the prereqs/coreqs/crosslisted sections on course cards. Also, I think if you change the column labels to "View on Canvas" and "Download PDF," it also makes sense to get rid of the "(Canvas Link)" and "(Download Link)" on each entry--I think it looks a bit cramped to have it everywhere. 

<img width="449" alt="image" src="https://github.com/user-attachments/assets/929c46d1-c2e3-415c-b1a9-4aa91eb6b626" />

6. some stuff (might) be getting duplicated? if this is how it's in the database, that's fine for now, but probably double check to make sure things aren't being duplicated just to be safe

<img width="177" alt="image" src="https://github.com/user-attachments/assets/55e5cee0-eaf4-457b-b4a1-20483dc62fc3" />
